test/e2e/federated-service.go

```diff
@@ -0,0 +1,283 @@
+			cfg, err := ccfg.ClientConfig()
+			Expect(err).NotTo(HaveOccurred())
+
+			cfg.QPS = KubeAPIQPS
+			cfg.Burst = KubeAPIBurst
+			clset := release_1_3.NewForConfigOrDie(restclient.AddUserAgent(cfg, UserAgentName))
+			clusterClientSets = append(clusterClientSets, clset)
+		}
+	})
+
+	AfterEach(func() {
+		framework.SkipUnlessFederated(f.Client)
+
+		// Delete the registered clusters in the federation API server.
+		clusterList, err := f.FederationClientset.Federation().Clusters().List(api.ListOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		for _, cluster := range clusterList.Items {
+			err := f.FederationClientset.Federation().Clusters().Delete(cluster.Name, &api.DeleteOptions{})
+			Expect(err).NotTo(HaveOccurred())
+		}
+	})
+
+	It("should be able to discover a federated service", func() {
+		framework.SkipUnlessFederated(f.Client)
+
+		createService(f.FederationClientset, clusterClientSets, f.Namespace.Name)
+
+		svcDNSNames := []string{
+			FederatedServiceName,
+			fmt.Sprintf("%s.%s", FederatedServiceName, f.Namespace.Name),
+			fmt.Sprintf("%s.%s.svc.cluster.local.", FederatedServiceName, f.Namespace.Name),
+			fmt.Sprintf("%s.%s.%s", FederatedServiceName, f.Namespace.Name, federationName),
+			fmt.Sprintf("%s.%s.%s.svc.cluster.local.", FederatedServiceName, f.Namespace.Name, federationName),
+		}
+		for _, name := range svcDNSNames {
+			discoverService(f, name, true)
+		}
+	})
+
+	It("should be able to discover a non-local federated service", func() {
+		framework.SkipUnlessFederated(f.Client)
+
+		createService(f.FederationClientset, clusterClientSets, f.Namespace.Name)
+
+		// Delete a federated service shard in the default e2e Kubernetes cluster.
+		err := f.Clientset_1_3.Core().Services(f.Namespace.Name).Delete(FederatedServiceName, &api.DeleteOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		waitForFederatedServiceShard(f.Clientset_1_3, f.Namespace.Name, nil, 0)
+
+		localSvcDNSNames := []string{
+			FederatedServiceName,
+			fmt.Sprintf("%s.%s", FederatedServiceName, f.Namespace.Name),
+			fmt.Sprintf("%s.%s.svc.cluster.local.", FederatedServiceName, f.Namespace.Name),
+		}
+		for _, name := range localSvcDNSNames {
+			discoverService(f, name, false)
+		}
+
+		svcDNSNames := []string{
+			fmt.Sprintf("%s.%s.%s", FederatedServiceName, f.Namespace.Name, federationName),
+			fmt.Sprintf("%s.%s.%s.svc.cluster.local.", FederatedServiceName, f.Namespace.Name, federationName),
+		}
+		for _, name := range svcDNSNames {
+			discoverService(f, name, true)
+		}
+	})
+})
+
+// waitForFederatedServiceShard waits until the number of shards of a given federated
+// service reaches the expected value, i.e. numSvcs in the given individual Kubernetes
+// cluster. If the shard count, i.e. numSvcs is expected to be at least one, then
+// it also checks if the first shard's name and spec matches that of the given service.
+func waitForFederatedServiceShard(cs *release_1_3.Clientset, namespace string, service *api.Service, numSvcs int) {
```

Comment:
nit: I don't understand how one could ever have more than 1 instance of the service in a given cluster at the same time.  There would be a name clash, right?  In which case, is the numSvcs parameter essentially a boolean?  Would it ever make sense to call this function with that parameter not equal to either 0 or 1?